sklearn/mixture/base.py

```diff
@@ -0,0 +1,441 @@
+
+        # Check all the parameters values of the derived class
+        self._check_parameters(X)
+
+    @abstractmethod
+    def _check_parameters(self, X):
+        """Check initial parameters of the derived class.
+
+        Parameters
+        ----------
+        X : array-like, shape  (n_samples, n_features)
+        """
+        pass
+
+    def _initialize_parameters(self, X):
+        """Initialize the model parameters.
+
+        Parameters
+        ----------
+        X : array-like, shape  (n_samples, n_features)
+        """
+        n_samples = X.shape[0]
+        random_state = check_random_state(self.random_state)
+
+        if self.init_params == 'kmeans':
+            resp = np.zeros((n_samples, self.n_components))
+            label = cluster.KMeans(n_clusters=self.n_components, n_init=1,
+                                   random_state=random_state).fit(X).labels_
+            resp[np.arange(n_samples), label] = 1
+        elif self.init_params == 'random':
+            resp = random_state.rand(X.shape[0], self.n_components)
+            resp /= resp.sum(axis=1)[:, np.newaxis]
+        else:
+            raise ValueError("Unimplemented initialization method '%s'"
+                             % self.init_params)
+
+        self._initialize(X, resp)
+
+    @abstractmethod
+    def _initialize(self, X, resp):
+        """Initialize the model parameters of the derived class.
+
+        Parameters
+        ----------
+        X : array-like, shape  (n_samples, n_features)
+
+        resp : array-like, shape (n_samples, n_components)
+        """
+        pass
+
+    def fit(self, X, y=None):
+        """Estimate model parameters with the EM algorithm.
+
+        The method fit the model `n_init` times and set the parameters with
+        which the model has the largest likelihood or lower bound. Within each
+        trial, the method iterates between E-step and M-step for `max_iter`
+        times until the change of likelihood or lower bound is less than
+        `tol`, otherwise, a `ConvergenceWarning` is raised.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points. Each row
+            corresponds to a single data point.
+
+        Returns
+        -------
+        self
+        """
+        X = _check_X(X, self.n_components)
+        self._check_initial_parameters(X)
+
+        # if we enable warm_start, we will have a unique init
+        do_init = not(self.warm_start and hasattr(self, 'converged_'))
+        n_init = self.n_init if do_init else 1
+
+        max_log_likelihood = -np.infty
+        self.converged_ = False
+
+        for init in range(n_init):
+            # print verbose msg
+            self._print_verbose_msg_init_beg(init)
+
+            if do_init:
+                self._initialize_parameters(X)
+
+            current_log_likelihood = -np.infty
+
+            for n_iter in range(self.max_iter):
```

Comment:
when `warm_start` is `True` and `self.n_iter_` is defined we should only perform `self.max_iter - self.n_iter_` additional iterations to be consistent with the meaning of `max_iter` of other models.